subsys/net/lib/websocket/websocket.c

```diff
@@ -715,94 +724,130 @@ int websocket_send_msg(int ws_sock, const uint8_t *payload, size_t payload_len,
+		if (ctx->parser_remaining == 0) {
+			if (ctx->masked) {
+				ctx->masking_value = 0;
+				ctx->parser_remaining = 4;
+				ctx->parser_state = WEBSOCKET_PARSER_STATE_MASK;
+			} else {
+				ctx->parser_remaining = ctx->message_len;
+				ctx->parser_state = WEBSOCKET_PARSER_STATE_PAYLOAD;
+			}
+		}
 		break;
-	case 0x0900:
-		*message_type_flag |= WEBSOCKET_FLAG_PING;
+	case WEBSOCKET_PARSER_STATE_MASK:
+		ctx->parser_remaining--;
+		ctx->masking_value |= (data << (ctx->parser_remaining * 8));
+		if (ctx->parser_remaining == 0) {
+			if (ctx->message_len == 0) {
+				ctx->parser_remaining = 0;
+				ctx->parser_state = WEBSOCKET_PARSER_STATE_OPCODE;
+			} else {
+				ctx->parser_remaining = ctx->message_len;
+				ctx->parser_state = WEBSOCKET_PARSER_STATE_PAYLOAD;
+			}
+		}
 		break;
-	case 0x0A00:
-		*message_type_flag |= WEBSOCKET_FLAG_PONG;
+	case WEBSOCKET_PARSER_STATE_PAYLOAD:
+		/* unreal condition, but check for safe */
+		if (ctx->data_buf_count < ctx->data_buf_size) {
+			ctx->data_buf[ctx->data_buf_count++] = data;
+		}
+		ctx->parser_remaining--;
+		if (ctx->parser_remaining == 0) {
+			ctx->parser_remaining = 0;
+			ctx->parser_state = WEBSOCKET_PARSER_STATE_OPCODE;
+		}
 		break;
+	default:
+		return -EFAULT;
 	}
 
-	len = value & 0x007f;
-	if (len < 126) {
-		len_len = 0;
-		*message_length = len;
-	} else if (len == 126) {
-		len_len = 2;
-		*message_length = sys_get_be16(&buf[2]);
-	} else {
-		len_len = 8;
-		*message_length = sys_get_be64(&buf[2]);
-	}
-
-	/* Minimum websocket header is 2 bytes, header length might be
-	 * bigger depending on length field len.
-	 */
-	*header_len = MIN_HEADER_LEN + len_len;
-
-	if (buf_len >= *header_len) {
-		if (value & 0x0080) {
-			*masked = true;
-			*mask_value = sys_get_be32(&buf[2 + len_len]);
-			*header_len += 4;
-		} else {
-			*masked = false;
-		}
-
-		return true;
+#if (LOG_LEVEL >= LOG_LEVEL_DBG)
+	if ((ctx->parser_state == WEBSOCKET_PARSER_STATE_PAYLOAD)
+			|| ((ctx->parser_state == WEBSOCKET_PARSER_STATE_OPCODE)
+					&& (ctx->message_len == 0))) {
+		NET_DBG("[%p] %smasked, mask 0x%08x, type 0x%02x, msg %zd",
+				ctx, ctx->masked ? "" : "un", ctx->masked ? ctx->masking_value : 0,
+				ctx->message_type, (size_t)ctx->message_len);
 	}
+#endif
 
-	return false;
+	return 0;
 }
 
 int websocket_recv_msg(int ws_sock, uint8_t *buf, size_t buf_len,
 		       uint32_t *message_type, uint64_t *remaining, int32_t timeout)
 {
 	struct websocket_context *ctx;
-	size_t header_len = 0;
-	int recv_len = 0;
-	size_t can_copy, left;
 	int ret;
 	k_timeout_t tout = K_FOREVER;
 
 	if (timeout != SYS_FOREVER_MS) {
 		tout = K_MSEC(timeout);
 	}
 
-#if defined(CONFIG_NET_TEST)
-	/* Websocket unit test does not use socket layer but feeds
-	 * the data directly here when testing this function.
-	 */
-	struct test_data {
-		uint8_t *input_buf;
-		size_t input_len;
-		struct websocket_context *ctx;
-	};
+	if ((buf == NULL) || (buf_len == 0)) {
+		return -EINVAL;
+	}
 
+#if defined(CONFIG_NET_TEST)
```

Comment:
I'm not really a fan of this ifdef slicing the function with test code, perhaps we could change the approach later (not requesting change for this PR).

Reply:
I completely agree with you! It is bad practice, but i think too many changes need for refuse it.

Reply:
I am the person who introduced this hack in the first place. I was very reluctant to do the testing like this but it was the simplest and easiest way at the time to do it as I did not had too much time to implement the testing part. And of course this technical dept easily sticks like a glue after the implementation.